### PR TITLE
Rewrite expo-app-auth

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

`expo-app-auth` historically was created to abstract the native AppAuth packages from the `Facebook` and `Google` modules. This was a quick and simple binding that resulted in 2 new native methods: `authAsync` and `refreshAsync`. This abstraction also enabled us to use app-auth in a sandboxed environment (the Expo client). 
But there were a couple core issues with the API:

<details><summary>Problems</summary>

#### Auth

The authentication method combined two steps together auth, and code exchange. This is default on iOS, and emulated on Android. When combined, the end-user has far less control over the login process, the errors are also somewhat cryptic because you're not completely sure which step failed. 

Another problem with this method is that the code exchange is just a basic fetch command, but we implement it on iOS and Android. Doing this makes it hard to push fixes and updates. `expo-app-auth` is a really important package for larger groups who use Expo, so we need the flexibility to fix bugs faster.

#### Custom API

The API we use isn't very cryptic but it's also not as clear as it could be. Having an API that extends the Node API makes it far easier for people migrating to `expo-app-auth`.

The current API also bundles the service config and issuer with the rest of the config. This is problematic because it resulted in our native code for every function looking like this: 
```
if (service config) do method
else resolve service config then do method
```
[Related PR](https://github.com/expo/expo/pull/7188)

This can lead to multiple unneeded fetch commands to the issuer to resolve the service config.

#### Overlapping Native Code

Instead of installing a whole new package on iOS, Android, and web, we should try to reuse the existing primitives. This leads to smaller native binaries. 

In the rewrite, expo-app-auth makes use of `expo-crypto`, `expo-random`, and `expo-web-browser` (optionally `expo-secure-store`). This removes a lot of native code from regular builds, in time we can also remove native AppAuth from the clients. 

We can also remove the `expo-app-auth` part of the kernel in favor of the `expo-web-browser` version in time. 

It might also make sense to delete the native code from `expo-google-sign-in` and just use `expo-app-auth` internally like we do with `expo-google`.

#### Limited API

Currently, `expo-app-auth` is missing an API for registering, code exchange, auth without exchange, and ending sessions. Now users will have access to the entire AppAuth API via Expo.

We can now make many more modifications to the API for use with more services. 

This becomes extremely important in cases like [Github auth](https://github.com/FormidableLabs/react-native-app-auth/issues/194). Reddit is another example of [non-standard auth that requires customization](https://github.com/reddit-archive/reddit/wiki/oauth2#retrieving-the-access-token).

It also wasn't possible for users to warm up or cool down the android web browser, resulting in slower startup times, this can now be done via the `expo-web-browser` package.

Users cannot clear cookies when they revoke their token, this means it's much harder (if not impossible) to switch users: [see related issue](https://github.com/expo/expo/issues/6756#issuecomment-584520743).

#### Non-universal

The API is currently not universal and not very web friendly. For example, say you redirect away from your website during authentication when you return the `expo-app-auth` API won't know how to continue where you left off. The proposed API persists the auth state using `AsyncStorage` before redirecting making it more web friendly. 

#### Poorly tested

Plenty of features from `expo-app-auth` are not tested. This is mostly due to the nature of native authentication, but with more control over the API we can test things like PKCE, or fetch requests in our E2E tests.

</details>

# How

- Extend the standard Expo API for `expo-app-auth`: `authAsync`, `refreshAsync`, `revokeAsync`, etc...
- Delete all native AppAuth code.
- Extend the node API `@openid/appauth` to use Expo primitives when possible.
- Create a more complex test in native component list that demonstrates authenticating with a couple different providers (on iOS, Android, and web).
- Create E2E tests for things like Crypto hashes and Base64 encoding.
- [x] Create a long term fix for broken Github OAuth 
  - resolve https://github.com/expo/expo/pull/3828 
  - related https://github.com/expo/expo/issues/3826
- [x] Add web support to `expo-web-browser` for redirecting in a popup window on web. https://github.com/expo/expo/pull/7479
- [x] Fix android web browser redirect https://github.com/expo/expo/issues/6679
- [x] Pass full server error back. Related https://github.com/expo/expo/issues/5868
- [x] Cancel should return with error code -3 https://github.com/expo/expo/issues/6033#issuecomment-546495331 
- [x] Ensure PKCE works without a client secret provided. related https://github.com/expo/expo/issues/6091
- [x] [Android] Open authentication browser in the same activity using `expo-web-browser`. related https://github.com/expo/expo/issues/5488
- [x] [Android] Support bare-workflow. 
  - related https://github.com/expo/expo/issues/5615
  - related https://github.com/expo/expo/issues/6696
- [ ] [Android] Ensure expo-web-browser redirects back in the managed workflow. related https://github.com/expo/expo/issues/7160
- [x] Pass extras into URL. related https://github.com/expo/expo/issues/6721
  - repro https://snack.expo.io/@aihen/app-auth-crash2
- [x] Ability to disable PKCE (and ensure PKCE doesn't cause breakage). related https://github.com/expo/expo/issues/6360

# Documentation

`expo-app-auth` is a universal implementation of AppAuth-JS. Instead of using AppAuth-iOS and AppAuth-Android, Expo AppAuth reimplements the API using native primitives like WebBrowser, Crypto, and Random to create a secure authentication solution across iOS, Android, and web. 

### Setup

- [ ] Client setup / comparison
- [ ] Android native setup
- [ ] iOS native setup

### Usage

Each topic should demonstrate the Expo standard API along with the underlying AppAuth-Expo API.

- [ ] Fetching a service config
- [ ] Auth (with auto exchange)
- [ ] Code Exchange
- [ ] Refreshing
- [ ] Revoking
- [ ] Dynamic Client Registration
- [ ] Warming the Android browser
- [ ] Clearing cookies
- [ ] Creating a redirect link for web with React Navigation
- [ ] Securely storing data using `expo-secure-store`. related https://github.com/expo/expo/issues/7179

### More Info

- [ ] AppAuth vs AuthSession vs `react-native-app-auth`
- [ ] Common errors (invalid_grant)
- [ ] Security concerns with client secrets

# Test Plan

- Native component list AppAuth screen should work with multiple providers
  - [ ] Github
    - Test extras (`response_mode: "form_post"`) https://github.com/expo/expo/issues/6721
  - [ ] Reddit
  - [ ] Coinbase (issue with PKCE)
  - [ ] [Identity Server 4](https://demo.identityserver.io/)
  - [ ] Google
  - [ ] FitBit
    - related https://github.com/expo/expo/issues/5868
  - [ ] Okta
  - [ ] Azure
  - [ ] Microsoft online 
    - related https://github.com/expo/expo/issues/6245
    - demo issue repo https://github.com/dsgrillo/expo-app-auth-ios-issue/blob/master/App.js
  - [ ] Slack
  - [ ] Spotify
  - [ ] GitLab
    - related https://github.com/expo/expo/issues/7234
    - resolve https://github.com/expo/expo/pull/7301
  - [ ] Dropbox
    - related https://github.com/expo/expo/issues/6360
    - repro https://snack.expo.io/@lerosen/appauth-android-bug
- Test Suite should test parts of the API that don't require the WebBrowser to open.

### CODE REVIEW

**Comment by mrmuhammadali** [comment]
Thanks @EvanBacon, appreciate your efforts.

File: apps/native-component-list/src/screens/AppAuth/AppAuthUtils.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+  }, []);
+
+  return link;
+}
+
+export function useParentNotifier(url: string) {
+  const link = useLinking();
+
+  React.useEffect(() => {
+    console.log('url >>>', url);
```
Comment (EvanBacon) [inline apps/native-component-list/src/screens/AppAuth/AppAuthUtils.ts:170]:
```suggestion
```

**Comment by EvanBacon** [comment]
Closing in favor of the new expo-auth-session API. There were some very fundamental problems with `expo-app-auth` that made it too awkward to use on web. In time we will be deprecating `expo-app-auth` in favor of `expo-auth-session`.

**Comment by AshleyRudland** [comment]
Hi @EvanBacon - will this or the new `expo-auth-session` resolve this issue: https://github.com/expo/expo/issues/6245?

If so, please let me know, have a project that cannot be released due to this issue. Thanks in advance!

**Pull request closed without merging**